In [1]:
import numpy as np   
wh=np.random.random((4,3))
bh=np.random.random((3,))
wout=np.random.random((3,))
bout=np.random.random((1,))

bout[0]=round(bout[0],2)


for i in range(4):
    for j in range(3):
        wh[i,j]=round(wh[i,j],2)

for i in range(3):
    bh[i]=round(bh[i],2)
    wout[i]=round(wout[i],2)

print("wh",wh)
print("bh",bh)
print("wout",wout)
print("bout",bout)

wh [[ 0.32  0.67  0.42]
 [ 0.19  0.42  0.89]
 [ 0.55  0.82  0.94]
 [ 0.25  0.53  0.63]]
bh [ 0.51  0.37  0.38]
wout [ 0.62  0.64  0.68]
bout [ 0.03]


## Broad Steps

**Step 0**: Read input and output

||**x**||||wh|||bh|||Hidden_layer_input||Hidden_layer_actions||wout|bout|output||y|E|
|---|---|:---:|:---:|:---:|:--:|:---:|:--:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|0|1|0||||||||||||||||1||
|1|0|1|1||||||||||||||||1||
|0|1|0|1||||||||||||||||0||

**Step 1**: Initialize weights and biases with random values (There are methods to initialize weights and biases but for now initialize with random values)

||**x**||||wh|||bh|||Hidden_layer_input||Hidden_layer_actions||wout|bout|output||y|E|
|---|---|:---:|:---:|:---:|:--:|:---:|:--:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|0|1|0|0.04|0.43|0.55|0.78|0.63|0.86||||||0.02|0.67|||1||
|1|0|1|1|0.49|0.04|0.04|||||||||0.24||||1||
|0|1|0|1|0.31|0.75|0.73|||||||||0.19||||0||
|||||0.69|0.82|0.16|||||||||||||||

**Step 2**: Calculate hidden layer input:

**hidden_layer_input = matrix_dot_product(X,wh) + bh **

||**x**|||wh|||bh|||Hidden_layer_input|||Hidden_layer_actions||wout|bout|output||y|E|
|---|---|:---:|:---:|:---:|:--:|:---:|:--:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|0|1|0|0.04|0.43|0.55|0.78|0.63|0.86|1.13|1.81|2.14|||0.02|0.67|||1||
|1|0|1|1|0.49|0.04|0.04||||1.82|2.63|2.3|||0.24||||1||
|0|1|0|1|0.31|0.75|0.73||||1.96|1.49|1.06|||0.19||||0||
|||||0.69|0.82|0.16|||||||||||||||

**Step 3**: Perform non-linear transformation on hidden linear input

**hiddenlayer_activations = sigmoid(hidden_layer_input) **


||**x**||||wh|||bh|||Hidden_layer_input|||Hidden_layer_actions||wout|bout|output||y|E|
|---|---|:---:|:---:|:---:|:--:|:---:|:--:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|0|1|0|0.04|0.43|0.55|0.78|0.63|0.86|1.13|1.81|2.14|0.76|0.86|0.89|0.02|0.67|||1||
|1|0|1|1|0.49|0.04|0.04||||1.82|2.63|2.3|0.86|0.93|0.91|0.24||||1||
|0|1|0|1|0.31|0.75|0.73||||1.96|1.49|1.06|0.88|0.82|0.74|0.19||||0||
|||||0.69|0.82|0.16|||||||||||||||

**Step 4**: Perform linear and non-linear transformation of hidden layer activation at output layer

__output_layer_input = matrix_dot_product (hiddenlayer_activations * wout ) + bout output = sigmoid(output_layer_input)__

||**x**||||wh|||bh|||Hidden_layer_input|||Hidden_layer_actions||wout|bout|output||y|E|
|---|---|:---:|:---:|:---:|:--:|:---:|:--:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|0|1|0|0.04|0.43|0.55|0.78|0.63|0.86|1.13|1.81|2.14|0.76|0.86|0.89|0.02|0.67|0.74||1||
|1|0|1|1|0.49|0.04|0.04||||1.82|2.63|2.3|0.86|0.93|0.91|0.24||0.75||1||
|0|1|0|1|0.31|0.75|0.73||||1.96|1.49|1.06|0.88|0.82|0.74|0.19||0.74||0||
|||||0.69|0.82|0.16|||||||||||||||

**Step 5**: Calculate gradient of Error(E) at output layer

__E = y-output__

||**x**||||wh|||bh|||Hidden_layer_input|||Hidden_layer_actions||wout|bout|output||y|E|
|---|---|:---:|:---:|:---:|:--:|:---:|:--:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|0|1|0|0.04|0.43|0.55|0.78|0.63|0.86|1.13|1.81|2.14|0.76|0.86|0.89|0.02|0.67|0.74||1|0.26|
|1|0|1|1|0.49|0.04|0.04||||1.82|2.63|2.3|0.86|0.93|0.91|0.24||0.75||1|0.25|
|0|1|0|1|0.31|0.75|0.73||||1.96|1.49|1.06|0.88|0.82|0.74|0.19||0.74||0|-0.74|
|||||0.69|0.82|0.16|||||||||||||||

**Step 6**: Compute slope at output and hidden layer

Slope_output_layer= derivatives_sigmoid(output)

Slope_hidden_layer = derivatives_sigmoid(hiddenlayer_activations)

||**x**||||wh|||bh|||Hidden_layer_input|||Hidden_layer_actions||wout|bout|output||y|E|
|---|---|:---:|:---:|:---:|:--:|:---:|:--:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|0|1|0|0.04|0.43|0.55|0.78|0.63|0.86|1.13|1.81|2.14|0.76|0.86|0.89|0.02|0.67|0.74||1|0.26|
|1|0|1|1|0.49|0.04|0.04||||1.82|2.63|2.3|0.86|0.93|0.91|0.24||0.75||1|0.25|
|0|1|0|1|0.31|0.75|0.73||||1.96|1.49|1.06|0.88|0.82|0.74|0.19||0.74||0|-0.74|
|||||0.69|0.82|0.16|||||||||||||||

||Slope_Hidden_layer||                        |Slope_output|
|:---:|:---:|:---:|                     |:---:|
|0.18|.12|0.10|                         |0.1924|
|0.12|0.07|0.08|                        |0.1875|
|.11|0.15|0.19|                         |-0.5476|

**Step 7**: Compute delta at output layer

__d_output = E * slope_output_layer*lr__

__lr:1__

||**x**||||wh|||bh|||Hidden_layer_input|||Hidden_layer_actions||wout|bout|output||y|E|
|---|---|:---:|:---:|:---:|:--:|:---:|:--:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|0|1|0|0.04|0.43|0.55|0.78|0.63|0.86|1.13|1.81|2.14|0.76|0.86|0.89|0.02|0.67|0.74||1|0.26|
|1|0|1|1|0.49|0.04|0.04||||1.82|2.63|2.3|0.86|0.93|0.91|0.24||0.75||1|0.25|
|0|1|0|1|0.31|0.75|0.73||||1.96|1.49|1.06|0.88|0.82|0.74|0.19||0.74||0|-0.74|
|||||0.69|0.82|0.16|||||||||||||||

||Slope_Hidden_layer||                     |Slope_output||Delta_output||
|:---:|:---:|:---:|                    |:---:|:---:|:---:|:---:|
|0.18|.12|0.10|                         |0.1924||0.05||
|0.12|0.07|0.08|                        |0.1875||0.05||
|.11|0.15|0.19|                         |-0.5476||0.41||


**Step 8**: Calculate Error at hidden layer

__Error_at_hidden_layer = matrix_dot_product(d_output, wout.Transpose)__

||**x**||||wh|||bh|||Hidden_layer_input|||Hidden_layer_actions||wout|bout|output||y|E|
|---|---|:---:|:---:|:---:|:--:|:---:|:--:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|0|1|0|0.04|0.43|0.55|0.78|0.63|0.86|1.13|1.81|2.14|0.76|0.86|0.89|0.02|0.67|0.74||1|0.26|
|1|0|1|1|0.49|0.04|0.04||||1.82|2.63|2.3|0.86|0.93|0.91|0.24||0.75||1|0.25|
|0|1|0|1|0.31|0.75|0.73||||1.96|1.49|1.06|0.88|0.82|0.74|0.19||0.74||0|-0.74|
|||||0.69|0.82|0.16|||||||||||||||

||Slope_Hidden_layer|                     ||Slope_output||Delta_output|||Error_at_hidden_layer
|:---:|:---:|:---:|                 |:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|0.18|.12|0.10|                         |0.1924||0.05||0.001|0.012|0.0095|
|0.12|0.07|0.08|                        |0.1875||0.05||0.001|0.012|0.0095|
|.11|0.15|0.19|                         |-0.5476||0.41||0.0082|0.0984|0.0779|


**Step 9**: Compute delta at hidden layer

d_hiddenlayer = Error_at_hidden_layer * slope_hidden_layer


||**x**||||wh|||bh|||Hidden_layer_input|||Hidden_layer_actions||wout|bout|output||y|E|
|---|---|:---:|:---:|:---:|:--:|:---:|:--:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|0|1|0|0.04|0.43|0.55|0.78|0.63|0.86|1.13|1.81|2.14|0.76|0.86|0.89|0.02|0.67|0.74||1|0.26|
|1|0|1|1|0.49|0.04|0.04||||1.82|2.63|2.3|0.86|0.93|0.91|0.24||0.75||1|0.25|
|0|1|0|1|0.31|0.75|0.73||||1.96|1.49|1.06|0.88|0.82|0.74|0.19||0.74||0|-0.74|
|||||0.69|0.82|0.16|||||||||||||||

||Slope_Hidden_layer|                     ||Slope_output||Delta_output|||Error_at_hidden_layer
|:---:|:---:|:---:|                 |:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|0.18|.12|0.10|                         |0.1924||0.05||0.001|0.012|0.0095|
|0.12|0.07|0.08|                        |0.1875||0.05||0.001|0.012|0.0095|
|.11|0.15|0.19|                         |-0.5476||0.41||0.0082|0.0984|0.0779|

||Delta_hidden_layer||
|:---:|:---:|:---:|
|0.000|0.001|0.000|
|0.000|0.000|0.000|
|0.000|0.015|0.015|

**Step 10**: Update weight at both output and hidden layer

wout = wout + matrix_dot_product (hiddenlayer_activations.Transpose, d_output) * learning_rate
wh = wh+ matrix_dot_product (X.Transpose,d_hiddenlayer) * learning_rate

||**x**||||wh|||bh|||Hidden_layer_input|||Hidden_layer_actions||wout|bout|output||y|E|
|---|---|:---:|:---:|:---:|:--:|:---:|:--:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|0|1|0|0.04|0.43|0.55|0.78|0.63|0.86|1.13|1.81|2.14|0.76|0.86|0.89|0.06|0.67|0.74||1|0.26|
|1|0|1|1|0.49|0.04|0.04||||1.82|2.63|2.3|0.86|0.93|0.91|0.28||0.75||1|0.25|
|0|1|0|1|0.31|0.75|0.73||||1.96|1.49|1.06|0.88|0.82|0.74|0.23||0.74||0|-0.74|
|||||0.69|0.82|0.16|||||||||||||||

||Slope_Hidden_layer|                     ||Slope_output||Delta_output|||Error_at_hidden_layer
|:---:|:---:|:---:|                 |:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|0.18|.12|0.10|                         |0.1924||0.05||0.001|0.012|0.0095|
|0.12|0.07|0.08|                        |0.1875||0.05||0.001|0.012|0.0095|
|.11|0.15|0.19|                         |-0.5476||0.41||0.0082|0.0984|0.0779|

||Delta_hidden_layer|||Learning_rate||
|:---:|:---:|:---:|:---:|:---|:---|:---:|:---:|
|0.000|0.001|0.000||0.1||
|0.000|0.000|0.000|
|0.000|0.015|0.015|


**Step 11**: Update biases at both output and hidden layer

bh = bh + sum(d_hiddenlayer, axis=0) * learning_rate
bout = bout + sum(d_output, axis=0)*learning_rate

||**x**||||wh|||bh|||Hidden_layer_input|||Hidden_layer_actions||wout|bout|output||y|E|
|---|---|:---:|:---:|:---:|:--:|:---:|:--:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|0|1|0|0.04|0.43|0.55|0.78|0.63|0.86|1.13|1.81|2.14|0.76|0.86|0.89|0.06|0.72|0.74||1|0.26|
|1|0|1|1|0.49|0.04|0.04||||1.82|2.63|2.3|0.86|0.93|0.91|0.28||0.75||1|0.25|
|0|1|0|1|0.31|0.75|0.73||||1.96|1.49|1.06|0.88|0.82|0.74|0.23||0.74||0|-0.74|
|||||0.69|0.82|0.16|||||||||||||||

||Slope_Hidden_layer|                     ||Slope_output||Delta_output|||Error_at_hidden_layer
|:---:|:---:|:---:|                 |:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|0.18|.12|0.10|                         |0.1924||0.05||0.001|0.012|0.0095|
|0.12|0.07|0.08|                        |0.1875||0.05||0.001|0.012|0.0095|
|.11|0.15|0.19|                         |-0.5476||0.41||0.0082|0.0984|0.0779|

||Delta_hidden_layer|||Learning_rate||
|:---:|:---:|:---:|:---:|:---|:---|:---:|:---:|
|0.000|0.001|0.000||0.1||
|0.000|0.000|0.000|
|0.000|0.015|0.015|

